torch/autograd/profiler.py

```diff
@@ -397,16 +403,23 @@ def __missing__(self, key):
 def parse_cpu_trace(thread_records):
     next_id = 0
     start_record = None
+    cuda_records = {}
     functions = []
     record_stack = []
     string_table = StringTable()
     # '__start_profile' is not guarenteed to be first, so we must find it here
     for record in itertools.chain(*thread_records):
         if record.name() == '__start_profile':
             start_record = record
-            break
+        elif record.name() == '__cuda_start':
+            cuda_records[record.device()] = record
     assert start_record is not None
 
+    # we don't know the correlation between CUDA/CPU time,
+    # instead, we just make sure the corresponding cuda kernels are shifted
+    # in time so they happen _after_ their CPU time, which is not
+    # alway show they get recorded.
+    max_shift = 0
```

Comment:
CUDA profiling won't be helpful until we can align the timelines properly. Can't you just use the CUDA start event to correlate them? Are the clocks getting out of sync?

Reply:
Well I am already using it, and I think it is helpful. Not being able to align the timeline is annoying but I still get nice traces per-GPU. 

This code simply provides  a global shift to all the cuda events so that they don't violate causality.

There is another clock sync problem where cuda events appear much later than they should due to some initialization problems I haven't been able to track down yet. As long as you make sure the device is already been used before profiling this part doesn't happen.

Reply:
It doesn’t violate causality, but having aligned timelines is incredibly useful - seeing how far ahead the CPU is compared to the GPU is one of the most important metrics. There is a function to get the GPU time somewhere, I can dig it up

Reply:
cuptiDeviceGetTimeStamp?
http://docs.nvidia.com/cuda/cupti/group__CUPTI__EVENT__API.html#group__CUPTI__EVENT__API_1g581c6a513e81f9431854561f43c06f36

Reply:
I mentioned that function in one of our meetings, but we finally converged to a decision, that we're giving up alignment for now, because the perf is similar to `nvprof`. So if you only care about recording the kernel durations, and don't want to mess with waiting until `nvprof` exits, you use this mode. Otherwise, if you really want a full profile, and want to see how far ahead is the CPU, you just use nvprof and inspect that.

Reply:
But it's an improvement we might want to add in the future

Reply:
The problem with `cuptiDeviceGetTimeStamp` is that there is no way I know of to turn a cudaEvent_t back into a (absolute) device timestamp. Knowing the correlation between cpu time and cuda device time doesn't help since I don't know the correlation between an event record and the GPU timestamp. Even the cuptiAPI doesn't seem to report the record time for an event, it only seems to record kernel start/stop times.